File: menu.py
import requests
import csv
import sys
import time

TOKEN = "d2f28b0ab69842ac91412d448ad10680"

# we now have this stores.csv file from the javascript code
# it is our job to parse it and send a request to the DD api
# to get the iced coffee price from each store

def getPrice(id):
    res = requests.get(
        f"https://mapi-dun.prod.ddmprod.dunkindonuts.com/menu-management/v1/menus/v4/{id}",
        headers={
            "Authorization": f"Bearer {TOKEN}"
        }
    )
    
    # The second box of categories is the non seasonal items
    allCats = res.json()["categoryFamilies"][1]["categories"]
    for cat in allCats:
        # Find iced drinks in non-seasonal items
        if cat["displayName"] == "Iced Drinks":
            icedDrinks = cat
            for singleIcedDrink in icedDrinks["categories"][0]["subCategories"]:
                # find coffee under iced drinks
                if singleIcedDrink["displayName"] == "Coffee":
                    icedCoffee = singleIcedDrink
                    for icedCoffeeVarient in icedCoffee["menuItems"]:
                        # find orignal blend under coffee
                        if icedCoffeeVarient["displayName"] == "Original Blend Iced Coffee":
                            # this is the price for a small
                            return icedCoffeeVarient["price"]

# Open the CSV file
with open('data/stores.csv', mode='r', newline='') as file:
    csv_reader = csv.reader(file)
    
    # Iterate over each row in the CSV
    for row in csv_reader:
        # skip first row
        if row[0] == "id":
            continue
        
        price = getPrice(row[0])
        print(f"Price for {row[1]}: {price}")
        #time.sleep(3)
        
        

File: menu.js
const axios = require("axios");
const fs = require("fs");
const TOKEN = "d2f28b0ab69842ac91412d448ad10680";
const CSV_PATH = "data/index.csv";

// if it doesn't exit, then create the header of the CSV
if (!fs.existsSync(CSV_PATH)) {
  fs.writeFileSync(CSV_PATH, "store,price,address,url,geoJSON\n", {
    flag: "w",
  });
  console.log("index.csv has been created.");
}

function parseMenuData(menuData, metaData) {
  var allCategories = menuData["categoryFamilies"];
  var validCat = null;
  for (var i = 0; i < allCategories.length; i++) {
    if (allCategories[i]["displayName"] == "Categories") {
      validCat = allCategories[i];
      break;
    }
  }
  if (validCat == null) {
    console.log("invalid categories for store " + menuData["menuId"]);
  }

  var validDrinkCat = null;
  for (var i = 0; i < validCat.categories.length; i++) {
    if (validCat.categories[i]["displayName"] == "Iced Drinks") {
      validDrinkCat = validCat.categories[i];
      break;
    }
  }
  if (validDrinkCat == null) {
    console.log("invalid drink categories for store " + menuData["menuId"]);
  }

  var icedDrinks = validDrinkCat["categories"][0]["subCategories"];
  var icedCoffee = null;
  for (var i = 0; i < icedDrinks.length; i++) {
    if (icedDrinks[i]["displayName"] == "Coffee") {
      icedCoffee = icedDrinks[i];
      break;
    }
  }
  if (icedCoffee == null) {
    console.log(
      "invalid iced coffee categories for store " + menuData["menuId"]
    );
  }

  var originalIcedCoffee = null;
  for (var i = 0; i < icedCoffee["menuItems"].length; i++) {
    if (
      icedCoffee["menuItems"][i]["displayName"] == "Original Blend Iced Coffee"
    ) {
      originalIcedCoffee = icedCoffee["menuItems"][i];
    }
  }
  if (originalIcedCoffee == null) {
    console.log("invalid iced coffee listing for store " + menuData["menuId"]);
  }

  var price = originalIcedCoffee["price"];
  console.log(price);

  // Format: store,price,address,url,geoJSON
  res = `${metaData["storeID"]},${price},${metaData["price"]},${metaData["url"]},${metaData["lat"]},${metaData["lon"]}\n`
  return res
}

async function scrapeMenu(storeID, storeMetaData) {
    setTimeout(() => {
        // Code to be executed after the delay
        console.log('Waited 2 seconds');
      }, 2000);

    // Validate arguments before anything
    if (!storeMetaData || !storeMetaData.hasOwnProperty("address")) {
      console.log("Invalid arguments passed for store:", storeID);
      return false;
    }
  
    const ENDPOINT = `https://mapi-dun.prod.ddmprod.dunkindonuts.com/menu-management/v1/menus/v4/${storeID}`;
    const CONFIGURATION = {
      headers: {
        Authorization: `Bearer ${TOKEN}`,
      },
    };
  
    try {
      const response = await axios.get(ENDPOINT, CONFIGURATION);
      // parseMenuData is sync :)
      await parseMenuData(response.data, storeMetaData);
    } catch (error) {
      console.error("Error scraping menu for store", storeID, error);
    }
  }
  

module.exports = {
  scrapeMenu,
};
